A76-ring-hash-improvements.md

```diff
@@ -93,13 +93,22 @@ At pick time:
 then the request hash key will be set to this value. If `request_hash_header`
 contains multiple values, then values are joined with a comma `,` character
 between each value before hashing.
-- If `request_hash_header` is not empty, and the request has no value or an
-empty value associated with the header defined, then the picker will generate a
-random hash for it. The use of a random hash key will effectively sends the
-request to a random endpoint.
 - If `request_hash_header` is empty, then the request hash key will be based on
 the xDS hash policy in RDS, which keeps the existing LB configuration for ring
 hash working as before with xDS.
+- If `request_hash_header` is not empty, and the request has no value associated
+with the header or its value is empty, then the picker behave as follows: the
+picker will generate a random hash for the request. If the use of this random
+hash triggers a connection attempt (according to the rules described in [A42:
+Picker Behavior][A42-picker-behavior]), then before queuing the pick, the picker
+will scan forward searching for a subchannel in `READY` state. If it finds a
+subchannel in `READY` state, the picker returns it.
+
+This behavior ensures that a single request creates at most one new connection,
+and that missing header does not add extra latency in the common case where
+there is already at least one subchannel in `READY` state. It converges to
+picking a random host, since each request may create a new connection to a
+random endpoint.
 
```

Comment:
I think it would be helpful to provide updated pseudo-code for the picker here.  Based on the pseudo-code in A61, I think the updated code will look like this:

```
// Determine request hash.
using_random_hash = false;
if (config.request_hash_header.empty()) {
  request_hash = call_attributes.hash;
} else {
  header = headers.find(config.request_hash_header);
  if (header == null) {
    using_random_hash = true;
    request_hash = ComputeRandomHash();
  } else {
    request_hash = ComputeHash(header);
  }
}

// Do pick based on hash.
first_index = ring.FindIndexForHash(request_hash);
requested_connection = false;
for (i = 0; i < ring.size(); ++i) {
  index = (first_index + i) % ring.size();
  if (ring[index].state == READY) {
    return ring[index].picker->Pick(...);
  }
  if (requested_connection) continue;
  if (ring[index].state == IDLE) {
    ring[index].endpoint.TriggerConnectionAttemptInControlPlane();
    if (using_random_hash) {
      requested_connection = true;
      continue;
    }
    return PICK_QUEUE;
  }
  if (ring[index].state == CONNECTING) {
    return PICK_QUEUE;
  }
}
if (requested_connection) return PICK_QUEUE;
return PICK_FAIL;
```

Reply:
I added the algorithm.